clients/drcachesim/tests/reuse_distance_test.cpp

```diff
@@ -358,36 +377,128 @@ print_histogram_mult_1p2_test()
+    }
 
     // Look for a few key strings in the output.
     std::vector<std::string> expected_strings = {
         "Reuse distance histogram bin multiplier: 1.2",
-        "Distance [min-max]        Count  Percent  Cumulative",
-        "       0 -        0           2       1%       1%",
-        "       1 -        1           2       1%       2%",
-        "      80 -       97          36      18%      98%",
-        "      98 -       99           4       2%     100%",
+        "Distance [min-max]        Count  Percent  Cumulative  :"
+        "       Count  Percent  Cumulative",
+        "       0 -        0           2       1%       1%     :"
+        "           2       1%       1%",
+        "       1 -        1           2       1%       2%     :"
+        "           0       0%       1%",
+        "      80 -       97          36      18%      98%     :"
+        "          10       5%      25%",
+        "      98 -       99           4       2%     100%     :"
+        "           0       0%      25%",
     };
     bool test_passes = find_strings_in_stream(expected_strings, output_string);
     assert(test_passes);
 }
 
+// Test the split of "everything" and "data" reuse-distance histogram.
+void
+data_histogram_test()
+{
+    std::cerr << "data_histogram_test()\n";
+
+    constexpr uint32_t LINE_SIZE = 32;
+
+    constexpr uint32_t TEST_ADDRESS = 0x1000; // Arbitrary.
+    constexpr uint32_t TEST_DISTANCE_START = 50;
+    constexpr uint32_t TEST_DISTANCE_END = 2000;
+    constexpr uint32_t TEST_DISTANCE_INCREMENT = 75;
+
+    // Create a reuse_distance test object.
+    reuse_distance_knobs_t knobs;
+    knobs.line_size = LINE_SIZE;
+    knobs.report_histogram = true;
+    knobs.verbose = 0;
+    reuse_distance_test_t reuse_distance(knobs);
+
+    // Create address generator with a predictable access pattern.
+    address_generator_t agen(TEST_ADDRESS, LINE_SIZE);
+
+    int instruction_hits = 0;
+    int data_hits = 0;
+    // Simple function to decide if a given access should be TRACE_TYPE_INSTR.
+    auto UseInstrType = [&](int distance) {
+        return (distance / TEST_DISTANCE_INCREMENT % 3) == 0;
+    };
+    for (int tgt_dist = TEST_DISTANCE_START; tgt_dist < TEST_DISTANCE_END;
+         tgt_dist += TEST_DISTANCE_INCREMENT) {
+        if (TEST_VERBOSE(1))
+            std::cerr << "Testing reuse dist=" << tgt_dist << "\n";
+        auto memref_type = UseInstrType(tgt_dist) ? TRACE_TYPE_INSTR : TRACE_TYPE_READ;
+        bool success =
+            generate_target_distance_memrefs(tgt_dist, reuse_distance, agen, memref_type);
+        if (UseInstrType(tgt_dist)) {
+            ++instruction_hits;
+        } else {
+            ++data_hits;
+        }
+        assert(success);
+    }
+    if (TEST_VERBOSE(1)) {
+        reuse_distance.print_results();
+    }
+
+    assert(reuse_distance.get_shard_map().size() == 1);
+
+    auto *shard = reuse_distance.get_aggregated_results();
+
+    assert(shard->data_refs > data_hits);
```

Comment:
Because it's comparing the number of data references to the number of **hits**, and by construction not every reference is a hit (to get a distance of X, the helper function inserts X unique references between two references to the same address).